# Задание 1 (на сопоставление)

1. z-тест - A) Анализ различий между многими выборками.
   
Условия:
- Независимость выборок.
- Нормальность распределения данных.Если размер выборки достаточно большой (больше 30),
можно применять тест и при отклонениях от нормальности.
- Неизвестна дисперсия генеральной совокупности, но размер выборки велик(больше 30), в этом случае
мы используем выборочную дисперсию в качестве оценки дисперсии генеральной совокупности.

2. t-тест - B) проверка среднего выборки на конкретное значение при неизвестной дисперсии.
   
Условия:
- Если у нас размер выборки меньше 30 и не знаем дисперсию генеральной совокупности.
- Нормальность распределения данных. Если размер выборки достаточно большой (больше 30),
можно применять тест и при отклонениях от нормальности.
- Все точки данных должны быть независимыми. 
- Для каждой выборки отклонения должны быть равны.
- t-тест является аналогом z-теста для случая, когда дисперсия или стандартное отклонение 
выборки неизвестно и должно быть оценено на основании самой выборки.
  
3. F-тест - C) соответствие структуры данных гипотетической модели. Сравнивает две дисперсии, чтобы проверить, 
происходят ли они из одной и той же совокупности.

Условия:
- Нормальность распределения данных. Если размер выборки достаточно большой (больше 30),
можно применять тест и при отклонениях от нормальности.
- Независимость выборок.
- Дисперсии в разных группах или выборках одинаковые. Однако возможно наличие двух идентичных средних значений
для двух разных выборок/групп, в то время как их дисперсии могут сильно различаться.

4. χ2-тест - D) проверка дисперсии выборки на конкретное значение при известном среднем.
   
Условия:
 - Критерий является непараметрическим (нет изначального предположения, что данные распределены 
нормально) и может работать даже для категориальных признаков.
- Независимость наблюдений.
- Известное среднее значение.
- Для корректного применения χ²-теста рекомендуется, чтобы размер выборки был достаточно большим (минимум 30).
  
5. ANOVA - E) проверка среднего выборки на конкретное значение при известной дисперсии.
   
Условия:
- Количественный непрерывный тип данных, дискретные данные менее желательны. 
- Независимые между собой выборки. 
- Нормальное распределение признака в статистических совокупностях, из которых извлечены 
выборки. 
- Равенство дисперсий изучаемого признака в статистических совокупностях,
из которых извлечены выборки (проверяется с помощью критерия Levene). 
- Независимые наблюдения в каждой из выборок.

6. F-тест - F) Сравнение дисперсий двух выборок.
   
Условия:
- Нормальность распределения данных. Если размер выборки достаточно большой (больше 30),
можно применять тест и при отклонениях от нормальности.
- Независимость выборок.
- Дисперсии в разных группах или выборках одинаковые. Однако возможно наличие двух идентичных средних значений
для двух разных выборок/групп, в то время как их дисперсии могут сильно различаться.

7. χ2-тест  G) Наличие или отсутствие связи между параметрами.

Условия:
- Категориальные данные.
- Критерий является непараметрическим (нет изначального предположения, что данные распределены 
нормально) и может работать даже для категориальных признаков.
- Независимость наблюдений.
- Известное среднее значение.
- Для корректного применения χ²-теста рекомендуется, чтобы размер выборки был достаточно большим (минимум 30).

# Задание 2 (про порошок)

Задача заключается в том, чтобы проверить, отклоняется ли средний вес порошка в коробке от установленного значения (3кг).
У нас есть информация о стандартном отклонении генеральной совокупности, и выборка достаточно велика.
Поэтому, для проверки гипотезы о среднем, следует использовать двусторонний z-тест. 

Формулировка гипотез:
- Нулевая гипотеза: Средний вес порошка в коробке не отличается от установленного значения. 
- Альтернативная гипотеза: Средний вес порошка в коробке отличается от установленного значения.

In [1]:
import numpy as np 
import scipy.stats as stats 
 
sample_mean = 3.005 
population_mean = 3 
population_std = 0.015 
sample_size = 50 
alpha = 0.05 
# Расчет z-оценки 
z_score = (sample_mean-population_mean)/(population_std/np.sqrt(sample_size)) 
print('z-оценка :',z_score) 
 
# Принятие решения на основе критического значения z 
# Расчет критической z-оценки 
z_critical = stats.norm.ppf(1-alpha/2) # Возвращает 95%-ный интервал значимости для двустороннего теста, от 2.5% до 97,5% 
print('Критическая z-оценка :',z_critical) 
if z_score >  z_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 
 
# Принятие решения на основе p-значения 
# P-Value : вероятность получить значение меньше, чем z-оценка, и больше, чем -z-оценка 
p_value = 1-stats.norm.cdf(z_score) + stats.norm.cdf(-z_score)  
print('p-value :',p_value) 
if p_value <  alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

z-оценка : 2.3570226039551083
Критическая z-оценка : 1.959963984540054
Отклонить нулевую гипотезу
p-value : 0.018422125454101483
Отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка меньше расчетного значения, 1.96 < 2.36, поэтому мы не можем принять 
нулевую гипотезу. 

Статистический вывод: на имеющихся данных p-value=0.0182, на уровне значимости 5% (уровне доверия 95%) есть 
основания отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: средний вес порошка в коробке отличается от установленного значения. Это подтверждается как результатом статистического теста с расчетом статистики z, так и p-value.

# Задание 3 (про двигатели)

Нам нужно использовать двусторонний несвязанный t-критерий для двух выборок с неравными дисперсиями, поскольку стандартные отклонения
генеральной совокупности неизвестны, размер выборки невелик, мы хотим выяснить наличие разницы
и в выборках представлены разные детали.

Формулировка гипотез:
- Нулевая гипотеза: Средние прочностные характеристики веществ одинаковы.
- Альтернативная гипотеза: Средние прочностные характеристики веществ различны.

In [2]:
import math
from scipy import stats

X1 = 90  # Средняя прочность вещества 1
X2 = 88  # Средняя прочность вещества 2
s1 = 2.3  # Стандартное отклонение вещества 1
s2 = 2.2  # Стандартное отклонение вещества 2
n1 = 15   # Размер выборки вещества 1
n2 = 10   # Размер выборки вещества 2

# Рассчитываем t-статистику
t_score = (X1 - X2) / math.sqrt((s1**2 / n1) + (s2**2 / n2))
print('t-оценка :',t_score) 

# Степени свободы
df = n1 + n2 - 2

# Принятие решения на основе критического значения t 
# Критическое значение t для уровня значимости 5% (двусторонний тест)
t_critical = stats.t.ppf(1 - 0.025, df)  # Для 5% уровня значимости, двусторонний тест
print('Критическая t-оценка :',t_critical) 
if t_score >  t_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

# Принятие решения на основе p-значения 
# p-value для рассчитанной t-статистики
p_value = 2 * (1 - stats.t.cdf(abs(t_score), df))  # Двусторонний тест 
print('p-value :',p_value) 
if p_value <  alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 


t-оценка : 2.186521551237011
Критическая t-оценка : 2.0686576104190406
Отклонить нулевую гипотезу
p-value : 0.03921015358894264
Отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка меньше расчетного значения, 2.07 < 2.19, поэтому мы не можем принять 
нулевую гипотезу. 

Статистический вывод: на имеющихся данных p-value=0.0329, на уровне значимости 5% (уровне доверия 95%) есть 
основания отвергнуть нулевую гипотезу в пользу альтернативы. 

Содержательный вывод: есть разница между средними прочностными характеристиками веществ. Это подтверждается как результатом статистического теста с расчетом статистики t, так и p-value.

# Задание 4 (про токи)

Будем использовать двусторонний несвязанный t-критерий для двух выборок с равными дисперсиями, поскольку стандартные отклонения
генеральной совокупности неизвестны, размер выборки невелик, мы хотим выяснить наличие разницы
и в выборках представлены тесты разных схем.

Формулировка гипотез:
- Нулевая гипотеза: Средний ток старой и новой схемы одинаков.
- Альтернативная гипотеза: Средний ток старой и новой схемы различен.

In [3]:
import numpy as np
from scipy import stats

# Токи старой и новой схемы
x1 = np.array([80.1, 82.3, 84.1, 82.6, 85.3, 81.3, 83.2, 81.7, 82.2, 81.4])
x2 = np.array([80.7, 81.3, 84.6, 81.7, 86.3, 84.3, 83.7, 84.7, 82.8, 84.4, 85.2, 84.9])

# Средние значения
mean_x1 = np.mean(x1)
mean_x2 = np.mean(x2)

# Дисперсия выборок
var_old = np.var(x1, ddof=1)
var_new = np.var(x2, ddof=1)

# Объединенная дисперсия
sp_squared = (((len(x1) - 1) * var_old) + ((len(x2) - 1) * var_new)) / (len(x1) + len(x2) - 2)
sp = np.sqrt(sp_squared)

# t-статистика
t_score = (mean_x1 - mean_x2) / (sp * np.sqrt(1/len(x1) + 1/len(x2)))
print('t-оценка :',t_score) 

# Степени свободы
df = len(x1) + len(x2) - 2

# Критическое значение t (для уровня значимости 5%)
alpha = 0.05
t_critical = stats.t.ppf(1 - alpha / 2, df)  # для двухстороннего теста

# Принятие решения на основе критического значения t 
t_critical = stats.t.ppf(1 - alpha/2, df) 
print('Критическая t-оценка:',t_critical) 
if np.abs(t_score) > t_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

# Принятие решения на основе p-значения 
# p-value для рассчитанной t-статистики
t_score, p_value = stats.ttest_ind(x1, x2) 
print('p-value :',p_value)
print('t-оценка :',t_score)
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

t-оценка : -1.8657246011714121
Критическая t-оценка: 2.0859634472658364
Нельзя отклонить нулевую гипотезу
p-value : 0.07681685091779608
t-оценка : -1.8657246011714124
Нельзя отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка больше модуля расчетного значения, 2.09 > 1.87, поэтому нет 
оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Статистический вывод: на имеющихся данных p-value=0.0768, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: Поскольку результаты статистического теста не показывают значимой разницы между токами старой и новой схемы, производителю можно перейти к внедрению новой схемы

# Задание 5 (про глюкозу)

Будем использовать парный двусторонний t-критерий для двух выборок, поскольку стандартные отклонения генеральной совокупности неизвестны, размер выборки невелик, а в выборках участвуют одни и те же люди

Формулировка гипотез:
- Нулевая гипотеза: Препарат не оказывает эффекта на уровень глюкозы, то есть разница между уровнями глюкозы до и после применения препарата равна нулю.
- Альтернативная гипотеза: Препарат оказывает эффект на уровень глюкозы, то есть разница между уровнями глюкозы до и после применения препарата не равна нулю.

In [4]:
import numpy as np 
import scipy.stats as stats 

# Данные
x1 = [9.6, 8.1, 8.8, 7.9, 9.2, 8.0, 8.4, 10.1, 7.8, 8.1]
x2 = [7.7, 6.2, 7.4, 7.5, 8.3, 6.2, 8.1, 8.9, 7.4, 7.0]
alpha = 0.05 
n = len(x1) 
x = np.zeros(0) 
for i in range (n): 
    x = np.append(x, x2[i]-x1[i]) 
d = x.mean() 
S = x.var()*n/(n-1) 
t_score = d*np.sqrt(n/S) 
print('t-score:', np.abs(t_score))

# Принятие решения на основе критического значения t 
t_critical = stats.t.ppf(1 - alpha/2, df = n-1) 
print('Critical t-Score:',t_critical) 
if np.abs(t_score) > t_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 
    
# Принятие решения на основе p-значения 
p_value = (1 - stats.t.cdf(np.abs(t_score), df = n-1)) * 2 
print('P-Value :',p_value) 
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

t-score: 5.71303238172197
Critical t-Score: 2.2621571628540993
Отклонить нулевую гипотезу
P-Value : 0.00028949403967537535
Отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка меньше расчетного значения, 2.62 < 5.71, поэтому мы не можем принять нулевую гипотезу. 

Статистический вывод: на имеющихся данных p-value=0.0003, на уровне значимости 5% (уровне доверия 95%) есть 
основания отвергнуть нулевую гипотезу в пользу альтернативы. 

Содержательный вывод: Препарат оказывает эффект на уровень глюкозы, то есть разница между уровнями глюкозы до и после применения препарата не равна нулю.

# Задание 6 (про тросы)

Будем использовать независимый односторонний t-критерий для двух выборок, приведенных степеней свободы. Поскольку выборки, из которых взяты образцы, не имеют равных дисперсий, размер выборок невелик, тесты проводились на разных кабелях и нужно выяснить, является ли новый кабель более устойчивым.

Формулировка гипотез:
- Нулевая гипотеза: Среднее напряжение разрушения нового кабеля не больше, чем у старого.
- Альтернативная гипотеза: Среднее напряжение разрушения нового кабеля больше, чем у старого.

In [5]:
import numpy as np
from scipy import stats

# Данные
x1 = np.array([90.1, 92.2, 94.3, 92.1, 93.4, 91.1, 93.2, 91.5])  # Старый кабель
x2 = np.array([92.7, 91.6, 94.7, 93.7, 96.5, 94.3, 93.7, 96.8, 98.9, 99.9])  # Новый кабель
alpha = 0.01

# Рассчитываем параметры для обеих выборок
mean_new = np.mean(x2)
mean_old = np.mean(x1)
std_new = np.std(x2, ddof=1)
std_old = np.std(x1, ddof=1)
n_new = len(x2)
n_old = len(x1)

# Стандартная ошибка разности средних
se = np.sqrt((std_new**2 / n_new) + (std_old**2 / n_old))

# t-статистика
t_score = (mean_new - mean_old) / se
print('t-score:', np.abs(t_score)) 

# Степени свободы для теста с неравными дисперсиями
df = ((std_new**2 / n_new + std_old**2 / n_old)**2) / \
     (( (std_new**2 / n_new)**2 / (n_new - 1)) + ((std_old**2 / n_old)**2 / (n_old - 1)))

# Принятие решения на основе критического значения t 
t_critical = stats.t.ppf(1 - alpha, df) # Критическое значение t для уровня значимости 0.01 и степеней свободы df
print('Critical t-Score:',t_critical) 
if np.abs(t_score) > t_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу")
    
# Принятие решения на основе p-значения 
# p-value для одностороннего теста (так как новый кабель должен быть прочнее)
p_value = 1 - stats.t.cdf(t_score, df)
print('P-Value :',p_value) 
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

t-score: 3.120397708249153
Critical t-Score: 2.6274728694881135
Отклонить нулевую гипотезу
P-Value : 0.003795023584113988
Отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка меньше расчетного значения, 2.63 < 3.12, поэтому мы не можем принять нулевую гипотезу.

Статистический вывод: на имеющихся данных p-value=0.0038, на уровне значимости 1% (уровне доверия 99%) есть основания отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: На основе результатов теста мы можем заключить, что новый кабель имеет статистически значимо большее напряжение разрушения, чем старый. Таким образом, изготовителю следует заменить старый кабель новым.

# Задание 7 (про абитуриентов)

Будем использовать парный двусторонний t-критерий для двух выборок, поскольку стандартные отклонения генеральной совокупности неизвестны, размер выборки невелик, а в выборках участвуют одни и те же люди

Формулировка гипотез:
- Нулевая гипотеза: Курсы не влияют на результаты, то есть разница между результатами до и после курса равна нулю.
- Альтернативная гипотеза: Курсы влияют на результаты, то есть разница между результатами до и после курса не равна нулю.

In [6]:
import numpy as np
import scipy.stats as stats

x1 = [7, 6, 5, 4, 6, 2, 10, 3, 8, 5] # Результаты до курсов
x2 = [9, 6, 4, 5, 7, 4, 10, 6, 9, 6] # Результаты после курсов
alpha = 0.01
n = len(x1)
x = np.zeros(0)
for i in range (n):
    x = np.append(x, x2[i]-x1[i])
d = x.mean()
S = x.var()*n/(n-1)
t_score = d*np.sqrt(n/S)
print('t-score:', np.abs(t_score))

# Принятие решения на основе критического значения t
t_critical = stats.t.ppf(1 - alpha/2, df = n-1)
print('Critical t-Score:',t_critical)
if np.abs(t_score) > t_critical:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")

# Принятие решения на основе p-значения
p_value = (1 - stats.t.cdf(np.abs(t_score), df = n-1)) * 2
print('P-Value :',p_value)
if p_value < alpha:
    print("Отклонить нулевую гипотезу")
else:
    print("Нельзя отклонить нулевую гипотезу")

t-score: 2.7386127875258306
Critical t-Score: 3.2498355415921254
Нельзя отклонить нулевую гипотезу
P-Value : 0.022899494551768207
Нельзя отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка больше модуля расчетного значения, 3.25 > 2.74, поэтому нет 
оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Статистический вывод: на имеющихся данных p-value=0.0229, на уровне значимости 1% (уровне доверия 99%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: Поскольку результаты статистического теста не показывают значимой разницы между результатами тестов абитуриентов до и после курсов, отклонить нуливую гипотезу нельзя, курсы не влияют на результат.

# Задание 8 (про фрукты и IQ)

Будем использовать  F-критерий, поскольку наша задача выяснить одинаковые ли дисперсии у обеих групп.

Формулировка гипотез:
- Нулевая гипотеза: Дисперсии выборок существенно не отличаются.
- Альтернативная гипотеза: Дисперсии выборок существенно отличаются. 

In [7]:
import scipy.stats as stats 
import numpy as np

# IQ людей, питавшихся апельсинами и грушами
x1 = np.array([125, 138, 112, 130, 124, 125, 110, 112, 117, 126]) 
x2 = np.array([124, 102, 142, 130, 131, 124, 127, 115, 108, 109]) 
alpha = 0.05 
n = len(x1) 
S1 = x1.var()*n/(n-1) 
S2 = x2.var()*n/(n-1) 
F_score = S1/S2 if S1 > S2 else S2 / S1
print('F-score:', np.abs(F_score))

# Принятие решения на основе критического значения F 
F_critical = stats.f.ppf(1 - alpha, dfn = n-1, dfd = n-1) 
print('Critical F-Score:', F_critical) 
if np.abs(F_score) > F_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу")
    
# Принятие решения на основе p-значения 
p_value = 1 - stats.f.cdf(np.abs(F_score), dfn = n-1, dfd = n-1) 
print('P-Value :',p_value) 
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

F-score: 1.9061769156692805
Critical F-Score: 3.178893104458269
Нельзя отклонить нулевую гипотезу
P-Value : 0.17529362058224196
Нельзя отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка больше модуля расчетного значения, 3.18 > 1.91, поэтому нет 
оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Статистический вывод: на имеющихся данных p-value=0.1753, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: Дисперсии IQ людей, питавшихся апельсинами и грушами, не различаются статистически значимо.

# Задание 9 (про самолеты)

Проверим, соответствует ли распределение Пуассона этим данным на уровне доверия 0.95

Формулировка гипотез:
- Нулевая гипотеза: Наши данные распределены по Пуассону.
- Альтернативная гипотеза: Наши данные не распределены по Пуассону.  

In [8]:
import numpy as np
from scipy.stats import poisson, chisquare, chi2

# Количество вылетов и соответствующие частоты
flights = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])
freq = np.array([36, 23, 11, 13, 8, 3, 2, 1, 3])

#  Считаем λ для распределения Пуассона)
total_flights = np.sum(flights * freq) 
total_planes = 100  
mean_flights = total_flights / total_planes

# Частоты для распределения Пуассона с параметром λ
freq_poisson = np.array([total_planes * poisson.pmf(k, mean_flights) for k in flights])
freq_poisson = freq_poisson * np.sum(freq) / np.sum(freq_poisson)

# Применяем тест хи-квадрат
res = chisquare(freq, freq_poisson) 
chi2_score, p_value = res.statistic, res.pvalue 
df = freq.shape[0]-1 
alpha = 0.05 
 
# Принятие решения на основе критического значения chi 
chi2_critical = chi2.ppf(1 - alpha, df)
print('chi2_score :',chi2_score)
print('chi2_critical :',chi2_critical)
if np.abs(chi2_score) > chi2_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 
 
# Принятие решения на основе p-значения
print('p_value :',p_value)
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу")

chi2_score : 72.50619032809553
chi2_critical : 15.50731305586545
Отклонить нулевую гипотезу
p_value : 1.5551461913805467e-12
Отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка меньше расчетного значения, 15.51 < 72.51, поэтому мы не можем принять нулевую гипотезу.

Статистический вывод: на имеющихся данных p-value почти равно нулю, на уровне значимости 5% (уровне доверия 95%) есть основания отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: На основе результатов теста мы можем заключить, что данные не распределены по Пуассону.

# Задание 10 (про цвет глаз и волос)

Для того чтобы определить, существует ли связь между цветом глаз и цветом волос, мы можем применить статистический тест на независимость хи-квадрат.

Формулировка гипотез:
- Нулевая гипотеза: Цвет глаз и цвет волос независимы.
- Альтернативная гипотеза: Цвет глаз и цвет волос зависимы. 

In [9]:
import numpy as np
from scipy.stats import chi2_contingency, chi2

# Наблюдаемые данные
X = np.array([[35, 52, 13],
              [28, 42, 10],
              [7, 10, 3]
])

res = chi2_contingency(X) 
chi2_score, p_value = res.statistic, res.pvalue 
nu = (X.shape[0]-1)*(X.shape[1]-1) 
alpha = 0.05 

# Принятие решения на основе критического значения chi2 
chi2_critical = chi2.ppf(1 - alpha, df=nu)
print('chi2_score :',chi2_score)
print('chi2_critical :',chi2_critical)
if np.abs(chi2_score) > chi2_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу")
    
# Принятие решения на основе p-значения 
print('p_value :',p_value)
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 


chi2_score : 0.09615384615384615
chi2_critical : 9.487729036781154
Нельзя отклонить нулевую гипотезу
p_value : 0.9988806869095861
Нельзя отклонить нулевую гипотезу


Выводы:

Статистический вывод: критическая оценка больше модуля расчетного значения, 9.49 > 0.096, поэтому нет 
оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Статистический вывод: на имеющихся данных p-value = 0.99888, на уровне значимости 5% (уровне доверия 95%) нет оснований отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: Связь между цветом глаз и цветом волос установить не удалось.

# Задание 11 (про тигров)

Задача заключается в том, чтобы оценить влияние типа корма на среднюю массу тигров. Для этого мы будем использовать однофакторный дисперсионный анализ (ANOVA). Если результат ANOVA покажет, что есть статистически значимые различия между группами, то мы применим тест Тьюки для выявления, какие именно группы отличаются друг от друга.

Формулировка гипотез:
- Нулевая гипотеза: Средние массы тигров для разных типов корма равны.
- Альтернативная гипотеза: Средние массы тигров для разных типов корма не равны. 

In [10]:
import numpy as np 
import pandas as pd
from statsmodels.stats.multicomp import pairwise_tukeyhsd
import scipy.stats as stats

# Данные о массе тигров для разных типов корма
X1 = np.array([151, 135, 137, 118, 132, 135, 131, 121, 137, 140, 152, 133, 151, 132])
X2 = np.array([108, 123, 145, 103, 110, 124, 116, 108, 104, 130, 106, 105, 114, 108])
X3 = np.array([140, 152, 149, 132, 124, 146, 156, 137, 124, 145, 147, 139, 157, 139])
X4 = np.array([127, 135, 110, 116, 119, 123, 102, 121, 118, 120, 114, 134, 121, 109])
df = pd.DataFrame() 
alpha = 0.05 
df['X1'] = X1 
df['X2'] = X2 
df['X3'] = X3
df['X4'] = X4 
X = np.array([df['X1'].mean(), df['X2'].mean(),df['X3'].mean(),df['X4'].mean()]) 
mean_all = X.mean() 
df_b = len(df.columns) - 1 
S_b = 0 
for i in X: 
    S_b = S_b + len(df)*(i - mean_all)**2 
S_b = S_b/df_b 
df_e = len(df.columns)*(len(df)-1) 
S_e = 0 
for i in df.columns: 
    df[i] = (df[i]-df[i].mean())**2 
    S_e = S_e + df[i].sum() 
S_e = S_e/df_e 
F_score = S_b/S_e 
print('F-Score:', F_score) 
 
# Принятие решения на основе критического значения F 
F_critical = stats.f.ppf(1 - alpha, dfn = df_b, dfd = df_e) 
print('Critical F-Score:', F_critical) 
if np.abs(F_score) > F_critical: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 
 
# Принятие решения на основе p-значения 
p_value = 1 - stats.f.cdf(np.abs(F_score), dfn = df_b, dfd = df_e) 
print('P-Value :',p_value) 
if p_value < alpha: 
    print("Отклонить нулевую гипотезу") 
else: 
    print("Нельзя отклонить нулевую гипотезу") 

# Если p-value < 0.05, то применяем тест Тьюки
if p_value < 0.05:
    print("\nЕсть статистически значимые различия между группами. Применяем тест Тьюки для парных сравнений.")
    data = np.concatenate([X1, X2, X3, X4])
    labels = ['Корм 1']*len(X1) + ['Корм 2']*len(X2) + ['Корм 3']*len(X3) + ['Корм 4']*len(X4)    
    # Используем Test Tukey для подробных сравнений между группами
    data_for_tukey = pd.DataFrame({"mass": data, "group": labels})
    tukey = pairwise_tukeyhsd(data_for_tukey['mass'], data_for_tukey['group'], alpha=0.05)
    print(tukey)
else:
    print("\nНет статистически значимых различий между группами.")

F-Score: 21.927813243053045
Critical F-Score: 2.782600423434607
Отклонить нулевую гипотезу
P-Value : 2.5927802127512223e-09
Отклонить нулевую гипотезу

Есть статистически значимые различия между группами. Применяем тест Тьюки для парных сравнений.
 Multiple Comparison of Means - Tukey HSD, FWER=0.05  
group1 group2 meandiff p-adj   lower    upper   reject
------------------------------------------------------
Корм 1 Корм 2    -21.5    0.0  -32.019  -10.981   True
Корм 1 Корм 3   5.8571  0.458  -4.6619  16.3762  False
Корм 1 Корм 4 -16.8571 0.0005 -27.3762  -6.3381   True
Корм 2 Корм 3  27.3571    0.0  16.8381  37.8762   True
Корм 2 Корм 4   4.6429 0.6473  -5.8762  15.1619  False
Корм 3 Корм 4 -22.7143    0.0 -33.2333 -12.1953   True
------------------------------------------------------


Выводы:

Статистический вывод: критическая оценка меньше расчетного значения, 2.78 < 21.93, поэтому мы не можем принять нулевую гипотезу.

Статистический вывод: на имеющихся данных p-value почти равно нулю, на уровне значимости 5% (уровне доверия 95%) есть основания отвергнуть нулевую гипотезу в пользу альтернативы.

Содержательный вывод: Существуют статистически значимые различия в массах тигров в зависимости от употребляемых ими кормов.

Из результатов теста Тьюки можно сделать выводы, что:
- Существуют значимые различия между типами корма №1 и №2, №1 и №4, №2 и №3, №3 и №4
- Различия между остальными типами крма не являются статистически значимыми.


# Задание 12 (про белок)

Для того чтобы оценить влияние факторов на массу тела самок африканских земляных белок с использованием метода последовательных сумм квадратов, нам нужно применить двухфакторный анализ дисперсии (ANOVA) с повторными измерениями,
так как у нас два фактора: возраст и выводок.

Формулировка гипотез:
- Нулевая гипотеза: Факторы (возраст и наличие выводка) не влияют на массу тела.
- Альтернативная гипотеза: Один или оба фактора влияют на массу тела.

In [11]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm

data = {
    'Возраст': ['1 год', '1 год', '1 год', '1 год', '1 год', 
                '2 года', '2 года', '2 года', '2 года', '2 года',
                '3 и более лет', '3 и более лет', '3 и более лет', '3 и более лет', '3 и более лет',
                '1 год', '1 год', '1 год', '1 год', '1 год', 
                '2 года', '2 года', '2 года', '2 года', '2 года',
                '3 и более лет', '3 и более лет', '3 и более лет', '3 и более лет', '3 и более лет'],
    'Выводок': ['Без выводка', 'Без выводка', 'Без выводка', 'Без выводка', 'Без выводка', 
                'Без выводка', 'Без выводка', 'Без выводка', 'Без выводка', 'Без выводка',
                'Без выводка', 'Без выводка', 'Без выводка', 'Без выводка', 'Без выводка',
                'С выводком', 'С выводком', 'С выводком', 'С выводком', 'С выводком',
                'С выводком', 'С выводком', 'С выводком', 'С выводком', 'С выводком',
                'С выводком', 'С выводком', 'С выводком', 'С выводком', 'С выводком'],
    'Масса': [440, 438, 429, 502, 602, 892, 868, 855, 866, 932, 
              1575, 849, 759, 1602, 1327, 308, 328, 326, 326, 325, 
              737, 798, 876, 810, 861, 1001, 901, 958, 1032, 883]
}

# Преобразуем в DataFrame
df = pd.DataFrame(data)

# Моделируем данные с помощью OLS
model = ols('Масса ~ C(Возраст) * C(Выводок)', data=df).fit()

# Выполним ANOVA (анализ дисперсии)
anova_results = anova_lm(model, typ=2)
print(anova_results)

                          sum_sq    df          F        PR(>F)
C(Возраст)             2427075.8   2.0  42.377707  1.333903e-08
C(Выводок)              202705.2   1.0   7.078626  1.368668e-02
C(Возраст):C(Выводок)    50688.2   2.0   0.885036  4.257428e-01
Residual                687269.6  24.0        NaN           NaN


Статистический вывод на основе p-value и уровня значимости 0.05:

- Фактор 'Возраст' влияет на массу тела (p-value = 0.0000)
- Фактор 'Выводок' влияет на массу тела (p-value = 0.0137)
- Взаимодействие 'Возраст' и 'Выводок' не влияет на массу тела (p-value = 0.4257)

Содержательный вывод: возраст и наличие выводка являются важными факторами, влияющими на массу тела самок африканских земляных белок. Также их взаимодействие показывает, что влияние каждого из этих факторов на массу тела является независимым, т.е. не зависит от сочетания этих двух факторов.